common/src/sphere_collider.rs

```diff
@@ -0,0 +1,800 @@
+) -> Option<f32> {
+    // This could be made more numerically stable by using a formula that depends on sinh_radius,
+    // but the precision requirements of collision should be pretty lax.
+    let mip_pos_a = math::mip(&ray.position, line_position);
+    let mip_dir_a = math::mip(&ray.direction, line_position);
+    let mip_pos_b = math::mip(&ray.position, line_direction);
+    let mip_dir_b = math::mip(&ray.direction, line_direction);
+
+    solve_quadratic(
+        mip_pos_a.powi(2) - mip_pos_b.powi(2) - cosh_radius.powi(2),
+        mip_pos_a * mip_dir_a - mip_pos_b * mip_dir_b,
+        mip_dir_a.powi(2) - mip_dir_b.powi(2) + cosh_radius.powi(2),
+    )
+}
+
+/// Finds the tanh of the distance a sphere will have to travel along a ray before it
+/// intersects the given point.
+fn solve_sphere_point_intersection(
+    ray: &Ray,
+    point_position: &na::Vector4<f32>,
+    cosh_radius: f32,
+) -> Option<f32> {
+    // This could be made more numerically stable by using a formula that depends on sinh_radius,
+    // but the precision requirements of collision should be pretty lax.
+    let mip_pos_a = math::mip(&ray.position, point_position);
+    let mip_dir_a = math::mip(&ray.direction, point_position);
+
+    solve_quadratic(
+        mip_pos_a.powi(2) - cosh_radius.powi(2),
+        mip_pos_a * mip_dir_a,
+        mip_dir_a.powi(2) + cosh_radius.powi(2),
+    )
+}
+
+/// Finds the lower solution `x` of `constant_term + 2 * half_linear_term * x + quadratic_term * x * x == 0`
+/// if such a solution exists. Assumes that `quadratic_term` is positive.
+///
+/// If a small perturbation to these terms would result in a solution of `x == 0.0`, this function has logic to
+/// to return Some(0.0) if three conditions hold in the context of collision checking:
+/// 1. The collider must be intersecting the object. This manifests as `constant_term <= 0.0`.
+/// 2. The collider must not be too far inside the object. This manifests as `constant_term >= -EPSILON`.
+/// 3. The direction of motion must be towards the collider. This manifests as `double_linear_term < 0.0`.
+fn solve_quadratic(constant_term: f32, half_linear_term: f32, quadratic_term: f32) -> Option<f32> {
+    const EPSILON: f32 = 1e-4;
+
+    // Extra logic to ensure precision issues don't allow a collider to clip through a surface
+    if (-EPSILON..=0.0).contains(&constant_term) && half_linear_term < 0.0 {
+        return Some(0.0);
+    }
+
+    let discriminant = half_linear_term * half_linear_term - quadratic_term * constant_term;
+    if discriminant < 0.0 {
+        return None;
+    }
+
+    // We use an alternative quadratic formula to ensure that we return a positive number if `constant_term > 0.0`.
+    // Otherwise, the edge case of a small positive `constant_term` could be mishandled.
+    // Note that discriminant can be negative, which allows this function to return NaN when there is no solution.
+    Some(constant_term / (-half_linear_term + discriminant.sqrt()))
```

Comment:
Can both `half_linear_term` and `discriminant` be zero here?

Reply:
Also, could `half_linear_term` be equal to `discriminant.sqrt()`?

Reply:
To get an infinity from this, `half_linear_term` has to be non-negative. This would only happen if the character is moving away from a wall, and no collision should be detected.

Infinity is outside the bounds of any range we choose, so this edge case shouldn't give us problems.

Reply:
I suppose infinity is a relatively sensible/harmless result.

Reply:
I definitely appreciate the call-out. Anywhere infinities can get involved, you can also end up with large numbers where you didn't intend them. I did need to think about it to be sure that we were fine after reading your question.

I think this is probably worth a code comment.

Reply:
Actually, thinking about this again, I think this does need to be addressed. If the numerator is also zero, we get NaN, and we cannot necessarily rule that out. I'll take a closer look at this.

Reply:
It looks like the answer is to restrict the solution space to non-negative lower solutions that are not double-roots (We don't care about double-roots because they correspond to moving tangent to a surface, which doesn't have to be a collision). This forces the constant term to be non-negative and the linear term to be negative. Because the linear term is negative, the denominator cannot be zero.